### ISSUE TITLE
KeyError / Attribute error when parsing OANDA JSON

### ISSUE THREAD

**Issue by jamespeterschinner** [jamespeterschinner/async_v20#12]
**Description of issue:**

When OANDA sends JSON object's. Occasionally the JSON data contains attributes that async_v20 wasn't expecting. This cause's a KeyError or AttributeError (depending if the JSON is in an array or not).

**Expected behaviour:**

async_v20 objects should contain all attributes that OANDA may send. 

**Actual behaviour:**

async_v20 doesn't contain all attributes OANDA's v20 API sends

**Explanation:**

There is currently no way to test for the completeness of async_v20's object model in respect to the actual data OANDA sends in the JSON response.  async_v20 has been constructed from OANDA's documentation which is not 100% perfect. OANDA has been contacted many times regarding this issue and I am awaiting there response. Where ideally the this issue can be resolved. By OANDA updating there documentation. 

Some sort of automated comparison between async_v20's object definitions and OANDA's documentation could then be set up.

The other option is to handle these errors in async_v20. Which I am reluctant to do because:

- Incomplete documentation OANDA should correct.
- Possibility of creating other errors in async_v20
- The ability of async_v20 to detect mistakes in user code may be compromised. Which may have real monetary cost if a mistake isn't caught early. 
- It would be a band aid of a larger problem

Having said all of that. You don't want your trading strategy to fail, due to an attribute in a JSON response that is likely irrelevant.

Until this issue can be resolved. I Encourage all users to post the stack trace of these errors here, so the async_v20's object model can be updated.

**Comment by jamespeterschinner** [comment]
Here are some example errors that have occurred in the past:
- 
![image](https://user-images.githubusercontent.com/31234531/33480350-afb03e76-d6e4-11e7-96f5-d29529ec5450.png)

**Comment by jamespeterschinner** [comment]
MarginUsed for TradeSummary is undocumented

![image](https://user-images.githubusercontent.com/31234531/33547388-0cff298e-d938-11e7-9ee6-ae4505b20759.png)

**Comment by jamespeterschinner** [comment]
MarginUsed  for TradeSummary is undocumented.


![image](https://user-images.githubusercontent.com/31234531/33547458-3a80b5d0-d938-11e7-965f-0a3387f116ab.png)

**Comment by jamespeterschinner** [comment]
- halfSpreadCost is undocumented for Transaction

![image](https://user-images.githubusercontent.com/31234531/33589793-052e6796-d9cf-11e7-8e96-bbaf32abe7c9.png)

**Comment by jamespeterschinner** [comment]
Version 2.2.4b4  contains halfSpreadCost. 

Required updating TradeReduce, TradeOpen & Transaction definitions

**Comment by jamespeterschinner** [comment]
Admittedly,  I'm getting frustrated about OANDA's undocumented changes to the v20 API, which keeps breaking async_v20. 

I need to set up some automated API stress test that can be run on a practice account. Where ideally all endpoints and responses get polled. Placing orders and the lot! There doesn't seem to be any other way to  check for consistency. :(

**Comment by jamespeterschinner** [comment]
CalculatedTradeState also has `margin_used` 

![image](https://user-images.githubusercontent.com/31234531/33749954-e837ff28-dc24-11e7-9f65-52083a9d7c92.png)

![image](https://user-images.githubusercontent.com/31234531/33749971-01e57644-dc25-11e7-96e9-9e4017e05c98.png)

**Comment by jamespeterschinner** [comment]
CalculatedPositionState has `margin_used`

![image](https://user-images.githubusercontent.com/31234531/33750118-cf3e99d6-dc25-11e7-8cbe-5997d0d36cb0.png)

![image](https://user-images.githubusercontent.com/31234531/33750128-dd1a2c8c-dc25-11e7-91cb-122246afb677.png)

**Comment by jamespeterschinner** [comment]
`partialFill` for Transaction is undocumented

![image](https://user-images.githubusercontent.com/31234531/33750369-a996fe1a-dc27-11e7-9e3a-93cd8de98776.png)

![image](https://user-images.githubusercontent.com/31234531/33750410-ea72b91a-dc27-11e7-9acf-ea386c7f763c.png)

**Comment by jamespeterschinner** [comment]
I have added [check_api_methods.py](https://github.com/jamespeterschinner/async_v20/blob/master/bin/check_api_methods.py) to catch these changes as they occur. 

Some endpoints need to be added, but it's a start

**Comment by jamespeterschinner** [comment]
OANDA has added:
 - `guaranteedStopLossOrderMode` to Account
 - `resettablePLTime` to Account
 - `guaranteedExecutionFees` to Account, Position & PositionSide

**Comment by jamespeterschinner** [comment]
Just added:

- `guaranteed` to Order.   (a bool) & corresponding version bump. `6.2.2b0` 
- `guaranteed` to Transaction.

**Comment by jamespeterschinner** [comment]
Ok, Well, I received confirmation that these extra attributes will be added to the documentation shortly. Standby!

**Comment by bjrnfrdnnd** [comment]
Hi there, 

is this project outdated?
I noticed that there are new undescribed features (e.g. guaranteed_stop_loss_order_mutability in accountSummary). 
I can repair these scalar values that are missing.

However, there seems to be an entire new class:
guaranteed_stop_loss_order_level_restriction

which is also found in the official v20 api:

```
class GuaranteedStopLossOrderLevelRestriction(BaseEntity):
    """
    A GuaranteedStopLossOrderLevelRestriction represents the total position
    size that can exist within a given price window for Trades with guaranteed
    Stop Loss Orders attached for a specific Instrument.
    """

    #
    # Format string used when generating a summary for this object
    #
    _summary_format = ""

    #
    # Format string used when generating a name for this object
    #
    _name_format = ""

    #
    # Property metadata for this object
    #
    _properties = spec_properties.primitives_GuaranteedStopLossOrderLevelRestriction

    def __init__(self, **kwargs):
        """
        Create a new GuaranteedStopLossOrderLevelRestriction instance
        """
        super(GuaranteedStopLossOrderLevelRestriction, self).__init__()
 
        #
        # Applies to Trades with a guaranteed Stop Loss Order attached for the
        # specified Instrument. This is the total allowed Trade volume that can
        # exist within the priceRange based on the trigger prices of the
        # guaranteed Stop Loss Orders.
        #
        self.volume = kwargs.get("volume")
 
        #
        # The price range the volume applies to. This value is in price units.
        #
        self.priceRange = kwargs.get("priceRange")

    @staticmethod
    def from_dict(data, ctx):
        """
        Instantiate a new GuaranteedStopLossOrderLevelRestriction from a dict
        (generally from loading a JSON response). The data used to instantiate
        the GuaranteedStopLossOrderLevelRestriction is a shallow copy of the
        dict passed in, with any complex child types instantiated
        appropriately.
        """

        data = data.copy()

        if data.get('volume') is not None:
            data['volume'] = ctx.convert_decimal_number(
                data.get('volume')
            )

        if data.get('priceRange') is not None:
            data['priceRange'] = ctx.convert_decimal_number(
                data.get('priceRange')
            )

        return GuaranteedStopLossOrderLevelRestriction(**data)
```

I am unable to solve this issue. I do not know which files and where I have to edit in async_v20 in order to include this new class.

Running ```python check_api_methods.py```
results in: 
```
async_v20.exceptions.InstantiationFailure: Could note create <class 'async_v20.definitions.types.Instrument'>. ARGUMENTS: ((), {'name': 'TRY_JPY', 'type': 'CURRENCY', 'display_name': 'TRY/JPY', 'pip_location': -2, 'display_precision': 3, 'trade_units_precision': 0, 'minimum_trade_size': '1', 'maximum_trailing_stop_distance': '100.000', 'minimum_trailing_stop_distance': '0.050', 'maximum_position_size': '0', 'maximum_order_units': '100000000', 'margin_rate': '0.05', 'minimum_guaranteed_stop_loss_distance': '0.200', 'guaranteed_stop_loss_order_execution_premium': '0.050', 'guaranteed_stop_loss_order_level_restriction': {'volume': '1000000', 'priceRange': '0.200'}}). __init__() got an unexpected keyword argument 'guaranteed_stop_loss_order_level_restriction'
Unclosed client session
```

**Comment by jamespeterschinner** [comment]
G'day mate,

Yes I haven't been maintaining this project in a while. I basically became
frustrated with the lack of documentation on OANDAs website for the v20
api, though maybe they have updated it.

Well done for finding the check_client_api_methods file. Just on my phone
now, I'll jump on my laptop and put together another email that should help
you Adrea's the issue.

**Comment by bjrnfrdnnd** [comment]
Thanks great!

I will try your solution. 
I was trying to reverse-engineer your code. 
I can file a pull request for the few scalar features I found (but that is not much).

I was also wondering what the descriptions.yaml file is doing. Code also works without it. Some sort of serialization thing, but I guess created when you dump some instances?

**Comment by jamespeterschinner** [comment]
OK, so the first thing I have done is hopped on the API documentation: http://developer.oanda.com/rest-live-v20/primitives-df/#InstrumentName

 



 

And yes I can see that there is this new class GuaranteededStopLossOrderLevelRestriction.

 

There are a few steps to add it (which I haven’t done in a while) so I’ll do my best here to show you what is required.

 

async_v20/definitions/attributes.py contain two python dictionaries {} that convert the CamelCase class attributes OANDA sends in it’s JSON response into snake_case attributes for the python class instance.

 

Refer to this piece of documentation: https://async-v20.readthedocs.io/en/latest/beyond_getting_started.html#what-might-be-useful

 



 

The new class contains two attributes ‘volume’ and ‘priceRange’ we need to make sure that both of these are in the two dictionaries contained within the attributes.py file. 

The first dictionary is instance_attributes this dictionary is used to parse CamelCase into snake_case. The ‘volume’ attribute is already in these dictionaries and we need not worry about it, that means that we need to add ‘priceRange’. See below: 



 

Note that I added ‘priceRange’ twice, one with the camelCase and key and another with the ‘snake_case’ key, this allows both camelCase and snake_case keyword arguments to be passed to the Class upon instantiation.

 

Now we need to add the reverse, so that the instance with snake_case attributes can be sent back to OANDA in camelCase. We do this in the json_attributes dictionary.

 



 

Ok, so we have added the attributes for the class but we still need to add the class. I have one this in async_v20/definitions/types.py.

Going by the error code you sent me in the first email it looks like this class is an attribute of an Instrument class, so I have defined this new class above the Instrument class definition. Notice that each argument has a type annotation and a default value. The ellipsis ‘…’ default value is always evaluated to True in python and I have used this to determine if OANDA sent the default value or not in a json response.

 



 

The next step is to add this new class to the Instrument type definition:

 



 

Now I have just realized that guaranteed_stop_loss_order_level_restriction is an attribute of an instrument which means that it needs to be added to the attributes.py dictionaries (instance_attributes and json_attributes). So you will have to repeat the above process converting from camelCase to snake_case.

 

 

I hope this has been useful for you. Let me know how you go, also regarding the repository I am no longer interested in maintaining this and would be willing to transfer it. Keep in mind that I didn’t add any documentation in the above process. I have had previous correspondence with OANDA regarding this issue see here: https://github.com/jamespeterschinner/async_v20/issues/12

 

Also I think that yaml file was a dump I performed when I did some web scraping originally to get the api specifications from OANDAs website which I used to ‘meta’ programme async_v20 in its very early stages.

**Comment by bjrnfrdnnd** [comment]
Many thanks. 

Most of your post is about what i referred to as "scalar information". This means adding stuff to the dictionaries in "attributes.py" and adding the keywords to the constructors in "types.py".

The part where I was uncertain was how to add the actual class. 
I noticed that you basically have an async_v20 class for each of Oanda's classes, but some classes in addition in primitives.py: for example "class TradeID". 

i do not understand the rationale behind how you definded the classes that you used; a Class from the official Oanda v20 repository can not directlly be copy-pasted to types.py; for example, some classes in types.py contain the custom class "TradeID" (e.g. TrailingStopLossOrderTransaction), some don't; (e.g. LimitOrder); and TradeID is not an Oanda-Class.

I do not understand the logic behind your class unverse vs Oanda's class universe. Therefore, I would be unable to add the new class to types.py because I do simply not know what to write in the class body. 

I guess it should be something like

```
class GuaranteedStopLossOrderLevelRestriction(BaseEntity):
    """
    A GuaranteedStopLossOrderLevelRestriction represents the total position
    size that can exist within a given price window for Trades with guaranteed
    Stop Loss Orders attached for a specific Instrument.
    """
    def __init__(self, volume: DecimalNumber = ..., priceRange: DecimalNumber = ...):
        Model.__init__(**locals())

```
But I am unsure about this. I am also unsure whether or not this class should be placed in primitives.py 
and not in types.py (it appears in the primitives section on the oanda REST API).

I also noticed that you have many more classes in your primitives.py as OANDA has; for example, TransactionFilter does not appear on OANDA and I do not know its purpose and the logic behind it.

I am certainly willing to continue this project but I fear I need a few more hints from you to understand the logic behind it.

**Comment by jamespeterschinner** [comment]
No worries,

I realized that you may have been viewing my email on github, in which case the screen shots I took, you wouldn’t have been able to see. 

Let me try and answer your questions. 

I looked into TradeID to refresh my memory regarding the design choice and found this comment I wrote.

# Technically OANDA's spec declares all specifiers as strings
# though TradeID and OrderID in async_v20 are integers. As this
# seems to be most useful type. We will make sure to cast them back
# to strings when sending JSON data to OANDA

So basically when specifying an order or trade OANDA expects to see a string in the JSON request. This string can be in the form of an integer eg. ‘1234’ ( async_v20:TradeID) or a user defined tag with the ‘@’ symbol prefix (async_v20:TradeSpecifier). But I wanted to treat the integer as an integer in async_v20 rather than a stringed representation of one. Hence I created an additional Class hirachy for this edge case, that allows both the TradeID and TradeSpecifier to be stored in there most intuitive form, but cast back to a string when sending the JSON request.

Regarding ‘unverse’, I think that might be a simple typo that I missed and should be fixed.


You guess at adding the Class is correct, but you will also have to modify the Instrument Class definition to include the this one you have added. 

Something like this:

‘’’guaranteed_stop_loss_order_level_restriction: GuaranteedStopLossOrderLevelRestriction = …’’’

Keep asking me questions and I will do my best to answer.

EDIT: I deleted my previous response and have re added it below to allow me to get the screenshots showing correctly

**Comment by jamespeterschinner** [comment]
OK, so the first thing I have done is hopped on the API documentation: http://developer.oanda.com/rest-live-v20/primitives-df/#InstrumentName

![1](https://user-images.githubusercontent.com/31234531/49704511-24dc0f00-fc68-11e8-8a8a-2c80bd67326a.jpg)

And yes I can see that there is this new class GuaranteededStopLossOrderLevelRestriction.

 
There are a few steps to add it (which I haven’t done in a while) so I’ll do my best here to show you what is required.

 

async_v20/definitions/attributes.py contain two python dictionaries {} that convert the CamelCase class attributes OANDA sends in it’s JSON response into snake_case attributes for the python class instance.

 

Refer to this piece of documentation: https://async-v20.readthedocs.io/en/latest/beyond_getting_started.html#what-might-be-useful
![2](https://user-images.githubusercontent.com/31234531/49704521-3fae8380-fc68-11e8-90cb-bd876acbea46.png)

 



 

The new class contains two attributes ‘volume’ and ‘priceRange’ we need to make sure that both of these are in the two dictionaries contained within the attributes.py file. 

The first dictionary is instance_attributes this dictionary is used to parse CamelCase into snake_case. The ‘volume’ attribute is already in these dictionaries and we need not worry about it, that means that we need to add ‘priceRange’. See below: 
![3](https://user-images.githubusercontent.com/31234531/49704530-52c15380-fc68-11e8-9f92-c268a6cf7964.jpg)

Note that I added ‘priceRange’ twice, one with the camelCase and key and another with the ‘snake_case’ key, this allows both camelCase and snake_case keyword arguments to be passed to the Class upon instantiation.

 
Now we need to add the reverse, so that the instance with snake_case attributes can be sent back to OANDA in camelCase. We do this in the json_attributes dictionary.

![4](https://user-images.githubusercontent.com/31234531/49704555-7f756b00-fc68-11e8-8e18-fd190aa3cdf6.jpg)


 

Ok, so we have added the attributes for the class but we still need to add the class. I have one this in async_v20/definitions/types.py.

Going by the error code you sent me in the first email it looks like this class is an attribute of an Instrument class, so I have defined this new class above the Instrument class definition. Notice that each argument has a type annotation and a default value. The ellipsis ‘…’ default value is always evaluated to True in python and I have used this to determine if OANDA sent the default value or not in a json response.


![5](https://user-images.githubusercontent.com/31234531/49704572-99af4900-fc68-11e8-9ae1-198eee001f8c.jpg)

 

The next step is to add this new class to the Instrument type definition:

 
![6](https://user-images.githubusercontent.com/31234531/49704576-adf34600-fc68-11e8-9f2e-beb661844e92.jpg)




 

Now I have just realized that guaranteed_stop_loss_order_level_restriction is an attribute of an instrument which means that it needs to be added to the attributes.py dictionaries (instance_attributes and json_attributes). So you will have to repeat the above process converting from camelCase to snake_case.

 

 

I hope this has been useful for you. Let me know how you go, also regarding the repository I am no longer interested in maintaining this and would be willing to transfer it. Keep in mind that I didn’t add any documentation in the above process. I have had previous correspondence with OANDA regarding this issue see here: https://github.com/jamespeterschinner/async_v20/issues/12

 

Also I think that yaml file was a dump I performed when I did some web scraping originally to get the api specifications from OANDAs website which I used to ‘meta’ programme async_v20 in its very early stages.

**Comment by bjrnfrdnnd** [comment]
Hi James, 

thanks again for the reply.
So you actually did send me screenshots?
I can not see them neither by email nor via github.

Anyways, I  forked your repo, created a new branch “updated-BN”,  committed my changes and pushed the whole thing to github.
I am new to this, so you might want to have a look.

I managed tweaking attributes.py, types.py and primitives.py such that your test code check_api_methods runs without errors. 
For this to happen, I added the new class to types.py and not to primitives.py, even if I do not understand how you actuallly decide where which class goes and
how exactly they are created.

If you have a second, could you explain to me why this new class goes into types.py and not into primitives.py?

Thanks, 
Bjoern

**Comment by bjrnfrdnnd** [comment]
Hi James, 

Now I can see the screenshots!

Thanks a lot!

Bjoern

**Comment by jamespeterschinner** [comment]
Great! good to hear, I'm sorry they didn't come through at the start, it would have been very confusing!

To answer your question regarding why is the new Class placed in `types.py` and not `primitives.py` is basically a design choice, I had a few iterations before I ended up with this structure. In previous versions of `async_v20` I had structured the modules and flies in the same manner as OANDA's online documentation, with the intention of creating an intuitive cross-reference system. This approach failed due to circular imports, that I could not resolve with out loading the whole Class hierarchy at once (This is the reason it was important you placed `GuaranteedStopLossOrderLevelRestriction` above the `Instrument` class definition. Having done that I realized that there was some structure i could preserve, which ended up being all the Classes in `primitives.py` essentially all classes in primitives.py are scalar values as you say. Either a string, float, integer or Pandas.Datetime, all other classes are built form combinations of these. The new `GuaranteedStopLossOrderLevelRestriction` class had two attributes and is there for not a single scalar value, which means it should be placed in `types.py` (according to my convention).

Another thing to keep in mind is that the docstring's for the classes are being used with Sphinx's autodoc allowing readthedocs to keep info updated, you may wish to keep these maintained as well.

**Comment by jamespeterschinner** [comment]
I have spent some time the last two days updating the object model incorporating changes. Some have been documented by OANDA and others have not. 

I have decided to issue warnings when unexpected attributes are supplied by OANDA rather than raise UnknownValue Exceptions. This will prevent async_v20 from breaking due to small changes from OANDA, but will mean that overtime greater drift between async_v20's object model and OANDA's may occur unless maintained properly.
